src/ring_theory/tensor_product.lean

```diff
@@ -1,36 +1,234 @@
+
+variables (R A M N)
+
+@[priority 500]
+instance restrict_scalars.semimodule : semimodule R (M ⊗[A] N) :=
+show semimodule R (restrict_scalars R A (M ⊗ N)), from
+restrict_scalars.semimodule R A (M ⊗ N)
+
+instance restrict_scalars.is_scalar_tower : is_scalar_tower R A (M ⊗[A] N) :=
+show is_scalar_tower R A (restrict_scalars R A (M ⊗ N)), from
+restrict_scalars.is_scalar_tower R A (M ⊗ N)
+
+/- Check that we didn't introduce a diamond. -/
+example : tensor_product.restrict_scalars.semimodule R R M N =
+          @tensor_product.semimodule R _ M N _ _ _ _ := rfl
+
+end restrict_scalars
+
+/-!
+### The `A`-module structure on `A ⊗[R] M`
+-/
+
+open linear_map
+open algebra (lsmul)
+
+section defn
+variables [comm_semiring R] [semiring A] [algebra R A]
+variables [add_comm_monoid M] [semimodule R M] [semimodule A M] [is_scalar_tower R A M]
+variables [add_comm_monoid N] [semimodule R N]
+
+variables (R A M N)
+
+@[priority 500]
+instance algebra_tensor_module : semimodule A (M ⊗[R] N) :=
+{ smul := λ a, (lsmul R M a).rtensor N,
+  one_smul := λ x, by simp only [alg_hom.map_one, one_def, id.def, id_coe, rtensor_id],
+  mul_smul := λ a b x, by simp only [alg_hom.map_mul, mul_def, rtensor_comp, comp_apply],
+  smul_add := λ a x y, by simp only [map_add],
+  smul_zero := λ a, by simp only [map_zero],
+  add_smul := λ a b x, by simp only [alg_hom.map_add, add_apply, rtensor_add],
+  zero_smul := λ x, by simp only [rtensor_zero, alg_hom.map_zero, zero_apply] }
+
+/- Check that we didn't introduce a diamond. -/
+example : tensor_product.algebra_tensor_module R R M N =
+          @tensor_product.semimodule R _ M N _ _ _ _ := rfl
+
+end defn
+
+namespace algebra_tensor_module
+
+section semiring
+variables [comm_semiring R] [semiring A] [algebra R A]
+variables [add_comm_monoid M] [semimodule R M] [semimodule A M] [is_scalar_tower R A M]
+variables [add_comm_monoid N] [semimodule R N]
+variables [add_comm_monoid P] [semimodule A P]
+
+variables {R A M}
+
+lemma smul_def (a : A) (x : M ⊗[R] N) : a • x = (lsmul R M a).rtensor N x := rfl
+
+@[ext] theorem ext (g h : (M ⊗[R] N) →ₗ[A] P)
+  (H : ∀ x y, g (x ⊗ₜ y) = h (x ⊗ₜ y)) : g = h :=
+linear_map.ext $ λ z, tensor_product.induction_on z (by simp_rw linear_map.map_zero) H $
+λ x y ihx ihy, by rw [g.map_add, h.map_add, ihx, ihy]
+
+variables (R A M N)
+
+instance is_scalar_tower :
+  is_scalar_tower R A (M ⊗[R] N) :=
+{ smul_assoc :=
+  begin
+    intros r a x,
+    simp only [algebra_tensor_module.smul_def, alg_hom.map_smul, smul_apply, rtensor_smul]
+  end }
+
+end semiring
+
+section comm_semiring
+variables [comm_semiring R] [comm_semiring A] [algebra R A]
+variables [add_comm_monoid M] [semimodule R M] [semimodule A M] [is_scalar_tower R A M]
+variables [add_comm_monoid N] [semimodule R N] [semimodule A N] [is_scalar_tower R A N]
+variables [add_comm_monoid P] [semimodule R P]
+
+def assoc : ((M ⊗[A] N) ⊗[R] P) ≃ₗ[A] (M ⊗[A] (N ⊗[R] P)) :=
+_
```

Comment:
I'm guessing this doesn't actually work?

Reply:
Done.